### PR TITLE
feat: improve editing experience

### LINKED ISSUE #43: Improve editing experience
Using this regularly I've noticed the edit and add item functionality isn't as efficient as it could be

Ideas
- Allow editing items inline, without a dialog
- Make checkboxes the default item type

### PR DESCRIPTION
**Pull request by Pertempto**
## What has changed

Replaced the EditItemDialog with a toolbar on the bottom of the list page.

## Connected Issues

Closes #43

## QA Steps

1. Open a list
2. Try to add/edit/delete items.
    - You should be able to do ALL the things you could do before with the edit item dialog.

### CODE REVIEW

**Comment by Pertempto** [comment]
I've got inline text editing to work! 🎉

**Comment by Pertempto** [comment]
# To Do
- [x] New item inline
- [x] Bottom toolbar
    - [x] Toggle type
    - [x] Repeat
        - [x] Implement a way to remove repeat config
    - [x] Delete item

**Comment by tycebrown** [comment]
Do you want me to merge `main` into this?

**Comment by Pertempto** [comment]
Don't do anything. I have it marked as draft as I make progress

**Comment by Pertempto** [comment]
I'll request review when it's ready

**Comment by Pertempto** [comment]
@tycebrown if you are curious to see what my goal is, try out my branch. Tap on an item to see the toolbar. Currently only the delete button works. I have to pause work until next week, because I will be camping this weekend.

**Comment by Pertempto** [comment]
@tycebrown It's been a long time since I used build runner. I tried to rename `RepeatConfiguration` to `RepeatConfig`, but I need help re-generating the `.g.` files. 


I tried using `flutter pub run build_runner build --delete-conflicting-outputs` but I get this error:

```
Unsupported type. Please annotate the property with @ignore.
package:lists/model/item_scheduling.dart:12:21
   ╷
12 │   late RepeatConfig repeatConfig;
   │                     ^^^^^^^^^^^^
```

Can you take a look? Do you know what I'm doing incorrectly?

**Comment by Pertempto** [comment]
It turns out I forgot to save some changes to `item_scheduling.dart` 🤦🏽 Build runner works now! 😄

**Comment by Pertempto** [comment]
@tycebrown This is ready for review! I added testing steps to the description. Be sure to try it out and check that my code looks good too!

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (Pertempto) [inline .github/PULL_REQUEST_TEMPLATE.md:1]: After we merge, github should automatically use this as a template for all our future PRs.

File: lib/model/repeat_config.dart
Comment (tycebrown) [inline lib/model/repeat_config.dart:1]: Like the rename!

File: lib/view/list_widget.dart [hunk 1]
```diff
@@ -101,34 +107,99 @@ class _ListWidgetState extends State<ListWidget> {
...
           {required Iterable<Item> fromItems, bool tappable = true}) =>
       fromItems
           .map((item) => ItemWidget(item,
               tappable: tappable,
-              onDelete: () async => await listModel.remove(item),
-              onEdited: () async => await listModel.update(item)))
+              onFocus: () => setState(() => selectedItem = item),
+              onUnfocus: () => print(
+                  'unfocus'), // TODO: figure out how to handle this better
+              // onUnfocus: () => setState(() => selectedItem = null),
```
Comment (tycebrown) [inline lib/view/list_widget.dart]: Do you think this MR (or issue) is the right place to address this to-do?

File: lib/view/item_widget.dart [hunk 2]
```diff
@@ -39,37 +68,52 @@ class _ItemWidgetState extends State<ItemWidget> {
           child: Checkbox(
             value: widget.item.isChecked,
             onChanged: _onNewCheckedState,
-            visualDensity: const VisualDensity(
-                horizontal: VisualDensity.maximumDensity,
-                vertical: VisualDensity.maximumDensity),
+            visualDensity: VisualDensity.compact,
```
Comment (tycebrown) [inline lib/view/item_widget.dart:67]: Looks much nicer

File: lib/view/item_widget.dart [hunk 3]
```diff
@@ -39,37 +68,52 @@ class _ItemWidgetState extends State<ItemWidget> {
...
+                    child: checkbox,
+                    visible: false,
+                    maintainSize: true,
+                    maintainAnimation: true,
+                    maintainState: true,
+                  ),
+                  _buildScheduledTimeStampChip(),
+                ],
+              )
+          ],
```
Comment (tycebrown) [inline lib/view/item_widget.dart]: This code hides the scheduling chip when the item's type is text (not checkbox). However, the user can still set the item to repeat:

https://github.com/Pertempto/Lists/assets/91034018/1e977647-a5d0-4c5f-b0a2-4568cd6c816a

I would propose instead either disabling the repeat button when the item's type is not checkbox or removing it from the toolbox. What do you think?

File: lib/view/list_widget.dart [hunk 4]
```diff
@@ -3,12 +3,17 @@ import 'dart:async';
 import 'package:collection/collection.dart';
 import 'package:lists/model/item.dart';
 import 'package:lists/model/list_model.dart';
-import 'package:flutter/material.dart';
+import 'package:flutter/material.dart' hide SearchBar;
 import 'package:lists/view/edit_item_dialog.dart';
```
Comment (tycebrown) [inline lib/view/list_widget.dart]:
```suggestion

```

File: lib/view/list_widget.dart [hunk 5]
```diff
@@ -3,12 +3,17 @@ import 'dart:async';
...
+import 'package:lists/view/repeat_dialog.dart';
 import 'package:lists/view/search_bar.dart';
 import 'package:lists/view/item_widget.dart';
+import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';
 import 'package:reorderables/reorderables.dart';
 
+import '../model/item_scheduling.dart';
+import 'confirmation_dialog.dart';
```
Comment (tycebrown) [inline lib/view/list_widget.dart]:
```suggestion

```

File: lib/view/list_widget.dart [hunk 6]
```diff
@@ -101,34 +107,99 @@ class _ListWidgetState extends State<ListWidget> {
...
+      child: Row(
+        children: item != null
+            ? [
+                if (item.itemType == ItemType.checkbox)
+                  IconButton(
+                    onPressed: () {
+                      item.itemType = ItemType.text;
+                      listModel.update(item);
+                    },
+                    icon: const Icon(Icons.text_format),
```
Comment (tycebrown) [inline lib/view/list_widget.dart]: Do you think [this icon](https://icon-sets.iconify.design/mdi/checkbox-blank-off-outline/) (`Mdi.checkboxBlankOffOutline`) would look better here?

**Review by tycebrown — CHANGES_REQUESTED** [review]
I really like the new way of editing items. However, there were two things I noticed:
- The user can add blank items:

https://github.com/Pertempto/Lists/assets/91034018/cb32094d-9c1f-460e-b6e5-2ce5d4a489da

This isn't a huge problem, but I think the fix is easy (on unfocus, just delete the item if it is completely empty).
- The biggest thing I noticed was that the user could still make edits to an item after it was deleted, filling the debug console with stack traces:

https://github.com/Pertempto/Lists/assets/91034018/2de09dd1-a57e-4495-a3d3-a3de37f7eed6

I haven't tested this yet, but one idea to fix this is to set the selected item to null when the user focuses away from an item.

File: lib/view/list_widget.dart (hunk 4 above)
Comment (tycebrown) [reply lib/view/list_widget.dart]: There was another old import in `item_widget.dart`, but I couldn't figure out how to add a suggestion there without making a commit (when I hovered over the line with the old import, there was no green 'plus' button).

File: lib/view/list_widget.dart (hunk 1 above)
Comment (Pertempto) [reply lib/view/list_widget.dart]: Oops, I forgot about this!

File: lib/view/list_widget.dart (hunk 6 above)
Comment (Pertempto) [reply lib/view/list_widget.dart]: Search here: https://pictogrammers.com/library/mdi/

Comment (Pertempto) [reply lib/view/list_widget.dart]: Done

File: lib/view/list_widget.dart (hunk 5 above)
Comment (Pertempto) [reply lib/view/list_widget.dart]: Done

File: lib/view/list_widget.dart (hunk 4 above)
Comment (Pertempto) [reply lib/view/list_widget.dart]: Done

File: lib/view/item_widget.dart (hunk 3 above)
Comment (Pertempto) [reply lib/view/item_widget.dart]: Fixed

**Comment by Pertempto** [comment]
> I haven't tested this yet, but one idea to fix this is to set the selected item to null when the user focuses away from an item.

That's what I originally tried, which you can see from my commented out code. It doesn't work however, because the selected item gets set to `null` BEFORE `_buildToolBar` is called, so the toolbar never functions correctly. I need to find a better way to handle losing focus.

File: lib/view/list_widget.dart (hunk 1 above)
Comment (Pertempto) [reply lib/view/list_widget.dart]: Done

**Comment by Pertempto** [comment]
I tried automatically deleting an item if it's empty, but that causes more issues that we don't really want to deal with.

It's common behavior for list apps to allow empty items, I don't see any problem with the current behavior. It's easy for users to delete the empty items if they want.

**Comment by Pertempto** [comment]
I think I've addressed all your concerns.

File: lib/view/item_widget.dart [hunk 7]
```diff
@@ -39,37 +64,57 @@ class _ItemWidgetState extends State<ItemWidget> {
...
+          const InputDecoration(border: InputBorder.none, isDense: true),
+      style: itemTextStyle.copyWith(decoration: textDecoration),
+      focusNode: _focusNode,
+      enabled: !widget.item.isChecked,
+    );
+
+    return GestureDetector(
+        child: Container(
+            // For some reason this makes the gesture detector work
+            color: Colors.transparent,
```
Comment (tycebrown) [inline lib/view/item_widget.dart:86]: Why do we need the gesture detector when the focus node (of the text field) already calls `widget.onFocus()` in its listener? I tried removing it, but I couldn't see any difference in functionality.

**Review by tycebrown — CHANGES_REQUESTED** [review]
One small thing: if the user is editing an item and decides to do a search, the toolbar for that item is still shown, even if that item is not in the search results:

https://github.com/Pertempto/Lists/assets/91034018/5449662b-3edf-470f-b0bc-b8e435a2edf4



Is this a problem? If so, is it worth fixing?

File: lib/view/item_widget.dart (hunk 7 above)
Comment (Pertempto) [reply lib/view/item_widget.dart:86]: you should be able to tap on ANY part of the item to select it, not just the text field.

**Comment by Pertempto** [comment]
> One small thing: if the user is editing an item and decides to do a search, the toolbar for that item is still shown, even if that item is not in the search results:
> 
> Is this a problem? If so, is it worth fixing?


I'll have to fix next week.

**Comment by Pertempto** [comment]
@tycebrown Try it now. I think it works as expected.

**Review by tycebrown — APPROVED** [review]
Works! 🎉

**Pull request merged**